posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,380 @@
+        return interval_sql
+
+    def retention_events_query(self):
+        _fields = [
+            self.get_timestamp_field(),
+            self.target_field(),
+        ]
+        params = {}
+
+        if self.event_query_type in [RetentionQueryType.TARGET, RetentionQueryType.TARGET_FIRST_TIME]:
+            source_timestamp = (
+                "min(e.timestamp)" if self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME else "e.timestamp"
+            )
+
+            _fields += [
+                f"""
+                [
+                    dateDiff(
+                        {{period}},
+                        {self.get_start_of_interval_sql(source='{start_date}')},
+                        {self.get_start_of_interval_sql(source=source_timestamp)}
+                    )
+                ] as breakdown_values
+                """
+            ]
+
+        params.update(
+            {
+                "start_date": self.query_date_range.date_from(),
+                "period": self.query_date_range.interval_name,
+            }
+        )
+
+        date_query, date_params = self._get_date_filter()
+        params.update(date_params)
+
+        filter_expressions = []
+        if self.query.properties is not None and self.query.properties != []:
+            filter_expressions.append(property_to_expr(self.query.properties, self.team))
+
+        filter_expressions.append(
+            self._get_entity_query(
+                entity=self.old_filter.target_entity
+                if self.event_query_type == RetentionQueryType.TARGET
+                or self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME
+                else self.old_filter.returning_entity
+            )
+        )
+
+        hogql_params = {key: ast.Constant(value=value) for key, value in params.items()}
+
+        data_query_hogql = parse_expr(date_query, placeholders=hogql_params)
+        if self.event_query_type != RetentionQueryType.TARGET_FIRST_TIME:
+            filter_expressions.append(data_query_hogql)
+
+        group_by_fields = None
+        having_expr = None
+        if self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            group_by_fields = [ast.Field(chain=["target"])]
+            having_expr = data_query_hogql
+        if self.event_query_type == RetentionQueryType.RETURNING:
+            group_by_fields = [ast.Field(chain=["target"]), ast.Field(chain=["event_date"])]
+
+        query = f"SELECT {','.join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}"
+        result = parse_select(
+            query,
+            placeholders={
+                **hogql_params,
+                "filter_where": ast.And(exprs=filter_expressions)
+                if len(filter_expressions) > 1
+                else filter_expressions[0],
+            },
+            timings=self.timings,
+        )
+        result.where = ast.And(exprs=filter_expressions)
+        result.group_by = group_by_fields
+        result.having = having_expr
```

Comment:
I'd opt for using `ast.Field` nodes instead of the `_fields` join here. I'd personally also add the where/group_by/having into the SQL statement directly and have these in the placeholders, e.g:
```sql
SELECT {fields}
FROM events
WHERE {where_clause}
GROUP BY {group_by_fields}
HAVING {having_expr}
```